### ISSUE TITLE
How samples are padded?

### ISSUE THREAD

**Issue by MagedSaeed** [kpu/kenlm#408]
Greetings @kpu 

Another question here with KenLM models. 

I am working on some experiments where the perplexity of statistical LMs (using KenLM) and Neural LMs (using GRUs) should be reported. What I noticed is that when considering the pad token in the training loss in the NLMs experiments, the results are significantly better than the statistical LMs, which is expected. However, the loss ignores the pad token, the results are worse than the statistical LMs. 

I reviewed your [paper](https://kheafield.com/papers/edinburgh/estimate_paper.pdf) and found out that samples are padded with `<s>` in the beginning, possibly, more than once. However, it is not clear when and how many padding tokens to add. Screenshot of this statement is below:

![image](https://user-images.githubusercontent.com/18549783/211073587-5a7502fd-6f88-49e2-9f12-6be5c68d1c0b.png)

Can you please elaborate on this? This turns out to be sensitive with perplexity as the frequency of these padding tokens affects the probabilities.

**Comment by kpu** [comment]
Multiple `<s>` padding is an implementation detail to count n-grams.  A sentence probability is always `p(A sentence . </s> | <s>)`

**Comment by MagedSaeed** [comment]
Thanks for your reploy and clarification
